tests/test_integration.py

```diff
@@ -0,0 +1,520 @@
+    result = run("riot --file rf.py", tmp_path)
+    assert (
+        result.stderr
+        == """
+Usage: riot [OPTIONS] COMMAND [ARGS]...
+Try 'riot --help' for help.
+
+Error: Invalid value for '-f' / '--file': Path 'rf.py' does not exist.
+""".lstrip()
+    )
+    assert result.returncode == 2
+
+    rf_path = tmp_path / "rf"
+    rf_path.write_text(
+        """
+from riot import Venv
+venv = Venv()
+""",
+    )
+    result = run("riot --file rf list", cwd=tmp_path)
+    assert (
+        result.stderr
+        == """
+Failed to construct config file:
+Invalid file format for riotfile. Expected file with .py extension got 'rf'.
+""".lstrip()
+    )
+    assert result.returncode == 1
+
+    rf_path = tmp_path / "riotfile.py"
+    rf_path.write_text(
+        """
+from riot import Venv
+venv = Venv()typo1234
+""",
+    )
+    result = run("riot --file riotfile.py list", cwd=tmp_path)
+    assert (
+        """
+Failed to construct config file:
+Failed to parse riotfile 'riotfile.py'.
+""".lstrip()
+        in result.stderr
+    )
+    assert (
+        """
+SyntaxError: invalid syntax
+""".lstrip()
+        in result.stderr
+    )
+    assert result.returncode == 1
+
+
+def test_help(tmp_path: pathlib.Path) -> None:
+    result = run("riot --help", cwd=tmp_path)
+    assert (
+        result.stdout
+        == """
+Usage: riot [OPTIONS] COMMAND [ARGS]...
+
+Options:
+  -f, --file PATH  [default: riotfile.py]
+  -v, --verbose
+  -d, --debug
+  --version        Show the version and exit.
+  --help           Show this message and exit.
+
+Commands:
+  generate  Generate base virtual environments.
+  list      List all virtual env instances matching a pattern.
+  run       Run virtualenv instances with names matching a pattern.
+""".lstrip()
+    )
+    assert result.stderr == ""
+    assert result.returncode == 0
+
+
+def test_version(tmp_path: pathlib.Path) -> None:
+    result = run("riot --version", cwd=tmp_path)
+    assert result.stdout.startswith("riot, version ")
+    assert result.stderr == ""
+    assert result.returncode == 0
+
+
+def test_list_no_file_empty_file(tmp_path: pathlib.Path) -> None:
+    result = run("riot list", cwd=tmp_path)
+    assert (
+        result.stderr
+        == """
+Usage: riot [OPTIONS] COMMAND [ARGS]...
+Try 'riot --help' for help.
+
+Error: Invalid value for '-f' / '--file': Path 'riotfile.py' does not exist.
+""".lstrip()
+    )
+    assert result.returncode == 2
+
+    rf_path = tmp_path / "riotfile.py"
+    rf_path.write_text(
+        """
+from riot import Venv
+""",
+    )
+    result = run("riot list", cwd=tmp_path)
+    assert result.stderr == ""
+    assert result.stdout == ""
+    assert result.returncode == 0
+
+
+def test_list_configurations(tmp_path: pathlib.Path) -> None:
+    rf_path = tmp_path / "riotfile.py"
+    rf_path.write_text(
+        """
+from riot import Venv
+venv = Venv(
+    name="test",
+    pys=[3],
+    command="echo hi",
+)
+""",
```

Comment:
I wonder if we should have a bunch of these as pytest fixtures, like:

```python
import pytest

@pytest.fixture
def riotfile_echo_hi(tmp_path: pathlib.Path) -> pathlib.Path:
    rf_path = tmp_path / "riotfile.py"
    rf_path.write_text(""" .... """)
    return rf_path
```


Could even consider the same for something like:

```python
@pytest.fixture
def run_riot(tmp_path: pathlib.Path) -> typing.Callable[[...] _T_CompletedProcess]:
    def wrapper(*args, **kwargs):
        return run(*args, **kwargs, cwd=tmp_path)
    return wrapper


def test_riot(run_riot) -> None:
    result = run_riot("riot list")
```


might be getting a bit carried away with these suggestions, but looking like a lot of repeated patterns in this test file.

Reply:
Yeah I was thinking about the second option because of the repeated `cwd=tmp_path` - I think it's worth doing

Reply:
Done in 7b2bb88. I'm personally ok with the boilerplate of creating a riotfile each time. If we need a common base one for a bunch of cases then we can create that fixture when needed.